source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+// SPDX-FileCopyrightText: Copyright (c) 2025 The Lethe Authors
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
+// core
+#include <core/grids.h>
+#include <core/lethe_grid_tools.h>
+
+#include <dem/find_cell_neighbors.h>
+#include <dem/insertion_file.h>
+#include <dem/insertion_list.h>
+#include <dem/ray_tracing.h>
+#include <dem/read_mesh.h>
+
+#include <sys/stat.h>
+
+
+template <int dim>
+RayTracingSolver<dim>::RayTracingSolver(
+  RayTracingSolverParameters<dim> parameters,
+  DEMSolverParameters<dim>        dem_parameters)
+  : mpi_communicator(MPI_COMM_WORLD)
+  , n_mpi_processes(Utilities::MPI::n_mpi_processes(mpi_communicator))
+  , this_mpi_process(Utilities::MPI::this_mpi_process(mpi_communicator))
+  , pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+  , parameters(parameters)
+  , dem_parameters(dem_parameters)
+  , triangulation(this->mpi_communicator)
+  , mapping(1)
+  , particle_handler(triangulation,
+                     mapping,
+                     DEMProperties::PropertiesIndex::n_properties)
+  , photon_handler(triangulation, mapping, dim)
+  , computing_timer(this->mpi_communicator,
+                    this->pcout,
+                    TimerOutput::summary,
+                    TimerOutput::wall_times)
+  , background_dh(triangulation)
+  , photon_displacement_vector(
+      parameters.ray_tracing_info.photon_displacement_vector)
+{
+  if (parameters.model_parameters.load_balance_method ==
+        Parameters::Lagrangian::ModelParameters<
+          dim>::LoadBalanceMethod::dynamic_with_sparse_contacts ||
+      parameters.model_parameters.load_balance_method ==
+        Parameters::Lagrangian::ModelParameters<
+          dim>::LoadBalanceMethod::dynamic)
+    throw std::runtime_error(
```

Comment:
Instead of throwing a raw assertion, please use the AssertThrow macro, I'm trying to clean it so that we use it everywhere.